File: 2026/January/Week1/SeMin/G5_7682.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	static char[][] board;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		StringBuilder sb = new StringBuilder();
		while(true) {
			String s = br.readLine();
			if(s.equals("end")) break;

			int xCnt = 0;
			int oCnt = 0;
			board = new char[3][3];
			for(int i = 0; i < 3; i++) {
				for(int j = 0; j < 3; j++) {
					board[i][j] = s.charAt(3 * i + j);

					if(board[i][j] == 'X') xCnt++;
					else if(board[i][j] == 'O') oCnt++;
				}
			}

			if( oCnt + xCnt == 9 && xCnt - oCnt == 1) {
				if(Check('X') && Check('O')) sb.append("invalid\n");
				else if(Check('O')) sb.append("invalid\n");
				else sb.append("valid\n");
			}else{ 
				if(Check('X') && Check('O')) sb.append("invalid\n");
				else if(Check('X') && xCnt-oCnt==1) sb.append("valid\n");
				else if(Check('O') && xCnt==oCnt) sb.append("valid\n");
				else sb.append("invalid\n");
			}
		}
		System.out.println(sb.toString());
	}
	
	private static boolean Check(char tiktakto) {
		for(int i = 0; i < 3; i++) {
			int cnt = 0;
			for(int j = 0; j < 3; j++) {
				if(board[i][j] == tiktakto) cnt++;
			}
			if(cnt == 3) return true;
		}

		for(int j = 0; j < 3; j++) {
			int cnt = 0;
			for(int i = 0; i < 3; i++) {
				if(board[i][j] == tiktakto) cnt++;
			}
			if(cnt==3) return true;
		}

		if(board[0][0] == tiktakto && board[1][1] == tiktakto && board[2][2] == tiktakto) return true;
		if(board[0][2] == tiktakto && board[1][1] == tiktakto && board[2][0] == tiktakto) return true;

		return false;
	}
}

File: January/Week1/SoHyun/G5_7682.py
def tickto(arr, t):
    if t == arr[0] == arr[1] == arr[2]:
        return True
    if t == arr[3] == arr[4] == arr[5]:
        return True
    if t == arr[6] == arr[7] == arr[8]:
        return True
    if t == arr[0] == arr[4] == arr[8]:
        return True
    if t == arr[2] == arr[4] == arr[6]:
        return True
    if t == arr[0] == arr[3] == arr[6]:
        return True
    if t == arr[1] == arr[4] == arr[7]:
        return True
    if t == arr[2] == arr[5] == arr[8]:
        return True 
    
    return False

while True:
    a = input()
    if a == "end":
        break
    else:
        flag = True
        arr = list(map(str, a))
        xcnt = arr.count('X')
        ocnt = arr.count('O')
        
        if xcnt > ocnt+1:
            print("invalid")
            continue
        
        if ocnt == xcnt:
            if tickto(arr, 'O') and not tickto(arr, 'X'):
                print("valid")
                continue
            
        if ocnt > xcnt:
            print("invalid")
            continue
        
        if ocnt + 1 == xcnt:
            if tickto(arr, 'X') and not tickto(arr, 'O'):
                print("valid")
                continue
            
        if xcnt == 5 and ocnt == 4:
            if not tickto(arr, 'O'):
                print("valid")
                continue
            
        print("invalid")